File: src/chess/game/pieces/Rook.java
package game.pieces;

// Import Java modules
import java.util.ArrayList;
import java.util.Collection;
import java.awt.Color;
import game.boardUtils.Square;
import game.utils.Constants;
import game.utils.FileC;
import game.utils.RankC;
import game.utils.UtilMethods;

public class Rook extends Piece {

    private ArrayList<Square> attackers;
    private boolean hasMoved;

    public Rook(FileC file, RankC rank, Color color) {
        super("Rook", file,rank,color);
        attackedPieces = new ArrayList<>();
        attackers = new ArrayList<>();
        hasMoved = false;
    }

    public ArrayList<Square> getAttackers() {
        return(attackers);
    }

    public boolean hasMoved() {
        return(hasMoved);
    }

    @Override
    public ArrayList<Square> getPossiblePositions(String pos, Collection<Square> squares) {
        ArrayList<Square> possiblePositions = new ArrayList<>(8);
        FileC file = FileC.valueOf(pos.substring(0, 1));
        RankC rank = RankC.valueOf(pos.substring(1));
        Square square = this.getSquare(squares, file, rank);
        if (square != null) {
            ArrayList<Square> ranks = square.getRanks(null);
            ArrayList<Square> files = square.getFiles(null);
            possiblePositions.addAll(ranks);
            possiblePositions.addAll(files);
            return(possiblePositions);
        }
        return(null);
    }

    @Override
    public boolean validate(FileC f, RankC r, Collection<Square> squares, King king) {
        int[] pos = UtilMethods.getWindowPos(this.getFile(), this.getRank(), Constants.SQUARE_WIDTH, Constants.SQUARE_HEIGHT);
        ArrayList<Square> possiblePositions = getPossiblePositions(UtilMethods.getGamePos(pos[0], pos[1], Constants.SQUARE_WIDTH, Constants.SQUARE_HEIGHT), squares);
        try{
            int moveRatio = (r.rank()-this.getRank().rank())/(((int) f.file())-((int) this.getFile().file()));
            if (moveRatio == 0) {
                boolean validated = false;
                int step = (((int) f.file())-((int) this.getFile().file()))/(Math.abs(((int) f.file())-((int) this.getFile().file())));
                for (int i = (int) this.getFile().file(); i != (int) f.file(); i+=step) {
                    FileC currFile = FileC.valueOf(Character.toString((char) i));
                    Square s = this.getSquare(squares, currFile, r);
                    if (s != null && !(s.getPiece() instanceof Empty)) {
                        validated = possiblePositions.contains(s);
                    } else {
                        validated = false;
                        break;
                    }
                }
                if (validated) {
                    Square startSquare = this.getSquare(squares, this.getFile(), this.getRank());
                    Square endSquare = this.getSquare(squares, f, r);
                    if (endSquare == null || endSquare.getPieceColor() == this.getColor()) {
                        validated = false;
                    } else {
                        if (possiblePositions.contains(endSquare)) {
                            Piece temp1 = startSquare.setPiece(endSquare.getPiece());
                            if (!(temp1 instanceof Empty)) {
                                endSquare.setPiece(new Empty(endSquare.getRank(), endSquare.getFile(), temp1.getColor()));
                            } else {
                                endSquare.setPiece(temp1);
                            }
                            validated = !king.check(king.getFile(), king.getRank(), squares);
                            endSquare.setPiece(startSquare.getPiece());
                            startSquare.setPiece(temp1);
                        } else {
                            validated = false;
                        }
                    }
                }
                return(validated);
            } else {
                return(false);
            }
        } catch (ArithmeticException e) {
            int step = (r.rank()-this.getRank().rank())/(Math.abs(r.rank()-this.getRank().rank()));
            boolean validated = false;
            for (int i=this.getRank().rank(); i != r.rank(); i+=step) {
                RankC currRank = RankC.valueOf(Integer.toString(i));
                Square s = this.getSquare(squares, f, currRank);
                if (s != null && !(s.getPiece() instanceof Empty)) {
                    validated = possiblePositions.contains(s);
                } else {
                    validated = false;
                    break;
                }
            }
            if (validated) {
                Square startSquare = this.getSquare(squares, this.getFile(), this.getRank());
                Square endSquare = this.getSquare(squares, f, r);
                if (endSquare == null || endSquare.getPieceColor() == this.getColor()) {
                    validated = false;
                } else {
                    if (possiblePositions.contains(endSquare)) {
                        Piece temp2 = endSquare.setPiece(startSquare.getPiece());
                        if (!(temp2 instanceof Empty)) {
                            endSquare.setPiece(new Empty(endSquare.getRank(), endSquare.getFile(), temp2.getColor()));
                        } else {
                            endSquare.setPiece(temp2);
                        }
                        validated = !king.check(king.getFile(), king.getRank(), squares);
                        startSquare.setPiece(endSquare.getPiece());
                        endSquare.setPiece(temp2);
                    } else {
                        validated = false;
                    }
                }
            }
            return(validated);
        }
    }

}

File: bin/chess/game/pieces/rook.py
from typing import Any
from utils.functions import get_window_pos, get_game_pos
import pygame
from .piece import Piece
from .king import King
from .empty import Empty
from utils.types import GamePosition, Squares


class Rook(Piece):

    def __init__(self, file: str, rank: int, color: pygame.Color, min_x: int, max_x: int, min_y: int, max_y: int, square_width: int, square_height: int, win_width: int, win_height: int) -> None:
        pygame.init()
        super().__init__(file, rank, 'Rook', color)
        self.min_x = min_x
        self.max_x = max_x
        self.min_y = min_y
        self.max_y = max_y
        self.square_width = square_width
        self.square_height = square_height
        self.win_width = win_width
        self.win_height = win_height
        self.x, self.y = self.piece_x, self.piece_y = get_window_pos(
            self.file, self.rank, self.possible_files, self.square_width, self.square_height)
        self.attacked_pieces = []
        self.has_moved = False

    def validate(self, position: GamePosition, squares: Squares, king: King) -> bool:
        file, rank = position
        try:
            move_ratio = (rank-self.rank)/(ord(file)-ord(self.file))
            if move_ratio == 0:
                squares[file+str(rank)], squares[self.file+str(self.rank)
                                                 ] = squares[self.file+str(self.rank)], squares[file+str(rank)]
                if king.check(squares=squares, position=position):
                    return False
                else:
                    stop = self.rank-rank
                    step = -1 if stop < 0 else 1
                    for i in range(0, stop, step):
                        compare_file = chr(ord(self.file)+i)
                        if not isinstance(squares[compare_file+str(self.rank+i)], Empty) and compare_file != file:
                            return False
                    else:
                        return True
        except ZeroDivisionError:
            stop = ord(self.file)-ord(file)
            step = -1 if stop < 0 else 1
            for i in range(0, stop, step):
                compare_rank = self.rank+i
                if not isinstance(squares[chr(ord(self.file)+i)+str(compare_rank)]) and compare_rank != rank:
                    return False
            else:
                return True
        return False

    # every val in matterial.values() must be an instance of Rook, Queen, Bishop, Pawn, King, or Knight
    def castle(self, king: King, matterial: dict[str, Any], squares: Squares) -> None:
        step = -1 if self.file < king.file else 1
        start = self.possible_files.index(king.file)
        files_to_check = self.possible_files[start::step][1::-1].sort()
        if len(files_to_check) == 3:
            third_check = files_to_check[2]
            third_square_empty_check = squares[f'{third_check}{self.rank}'].is_empty
        else:
            third_check = third_square_empty_check = True
        if squares[f'{files_to_check[0]}{self.file}'].is_empty and self.squares[f'{files_to_check[1]}{self.rank}'].is_empty and third_square_empty_check:
            if not (king.check(position=(files_to_check[0], self.rank), squares=squares) or king.check(position=(files_to_check[1], self.rank), squares=squares) or king.check(position=(third_check, self.rank), squares=squares) or king.check(position=(self.file, self.rank), squares=squares)):
                if not (self.has_moved or king.has_moved):
                    king.file = files_to_check[1] if len(
                        files_to_check) == 2 else files_to_check[1]
                    self.file = files_to_check[0] if len(
                        files_to_check) == 2 else files_to_check[2]
